### DISCUSSION TITLE
Overhaul equality-checking and type-coercion

### DISCUSSION THREAD

**Discussion by KingSupernova31** [godotengine/godot-proposals#10553] [Scripting]
The current type system in GDscript is kind of a mess.

It performs implicit type coercion on statically-typed variables, which defeats the purpose of having static typing:
```
var foo := false
foo = 17.5
print(foo) #true
```
It errors if you try to compare dynamically-typed variables of different types, which defeats the purpose of having dynamic typing:
```
var foo = true
print(foo == 17.5) #typeerror
```
It has inconsistent type-checking behavior for statically-typed variables:
```
var foo: bool
print(foo == 0) #typeerror
print(foo == null) #false
```

Various proposals have been made for new features that behave in more expected ways, like `is_same()` and `===`. My suggestion: learn from Javascript's mistakes, and don't go down that path.

Javascript's typing system is [famously](https://i.imgur.com/epeo0GS.jpg) bad. Many of these quirks are [not things you would ever do in a serious program](https://www.destroyallsoftware.com/talks/wat), but several of them are. I've personally, for example, been tripped up by the fact that `NaN != NaN`, but `[NaN].includes(NaN)`. Javascript has both a loose equality `==` operator and a strict equality `===` operator, but it is nigh-universally [considered bad practice](https://stackoverflow.com/questions/359494/which-equals-operator-vs-should-be-used-in-javascript-comparisons) to ever use `==`. (Unless you're comparing strings, because for some godforsaken reason `"foo" !== new String("foo")`, but `"foo" == new String("foo")`) It also needs yet more equality checks for things `===` can't handle, like `isNaN()` (which is itself [famously confusing](https://x.com/IsaacKing314/status/1737774263942213901)) and `Number.isNaN()` (which does something slightly different).

This sort of thing is bad design. Sure, you can learn the pitfalls and how to avoid them (or just use Typescript), but especially for beginners it can be very frustrating. Since GDscript is specifically aimed to be easy to learn, I think avoiding this sort of mish-mash of inconsistent behavior should be a priority.

Javascript's woes come largely from noticing problems with previous implementations, but not wanting to break existing websites, so they keep adding new things that are similar but subtly different from an existing feature. Godot has two big advantages here:

1. It's still quite new, and rapidly growing. This means that the number of existing users who would be adversely affected by a breaking change is small relative to the number of future users who will be helped. Much better to rip the band-aid off now than to spend years making the language more and more bloated with redundant features.
2. Users of Javascript are forced to use the latest version, since browsers update automatically, which means there's a very good reason for them to make everything backwards compatible. But users of Godot can just choose not to update and continue using an older version for as long as they want to, so breaking changes don't actually hurt existing Godot developers unless they want access to some other new feature.

I'm still pretty new to GDscript, so I don't have strong opinions on exactly how types _should_ work, just that the current system is confusing and error-prone. I would generally suggest that:

* Errors should be designed not only to cover "the compiler doesn't know what this means", but also "the dev obviously didn't mean to do this". If there's a line of code that there's no good reason someone would want to write, and this can be easily checked by the computer, it should be. (Could be a warning instead of an error if preferred.)
* Operations should be easier to type the more commonly used they are. Godot 4 introduced `is_same()` to solve most of the problems that `==` had. This is completely backwards; now instead of typing 2 characters, I have to type 10. (This is pretty much what Javascript did by making everyone use `===` instead of `==`, but 3 times longer.)
* If something should obviously work, make it work. Being unable to check if a variable equals something by writing `variable == thing` is incredibly unintuitive.
* Be willing to bite the bullet on having different types be always unequal or incommensurable. I understand that 1 == 1.0 makes sense from a pure math perspective, but A) this is inconsistent with the fact that other numerical operations do respect types, like how `5/2` is `2` rather than `2.5`, and B) it's bad practice to check floats for exact equality anyway, since floating point errors could make it differ by a tiny amount. Comparing ints to floats usually shouldn't be happening, so it seems fine to me if there's a slight additional friction to do so, like having to use `is_equal_approx()`.
* Special operators should only differ from the normal ones in the advertised way, and have consistent behavior otherwise. For example it makes sense to have `is_equal_approx()` available for use with floats, but why does `0 == false` result in an error, `is_same(0, false) == false`, yet `is_equal_approx(0, false) == true`? Nobody could reliably predict that. A function like `is_equal_approx()` should behave exactly like the base equality check, with the _only_ difference being its handling of floats, like it says in the name. (Or just error on inputs that are not numbers.)
* Be ok with a tiny performance loss if it leads to significantly better usability. Anyone who needs serious performance is probably going to be using C++ anyway.

**Comment by Calinou** [comment]
Moving to discussion, as there is no concrete technical implementation provided.

> It's still quite new, and rapidly growing. This means that the number of existing users who would be adversely affected by a breaking change is small relative to the number of future users who will be helped. Much better to rip the band-aid off now than to spend years making the language more and more bloated with redundant features.

This is much less true since Godot 4.0's release. Compatibility-breaking changes where no automated upgrade path can be provided are now avoided as much as possible.

Anything that makes the type system stricter should be a warning by default. GDScript's warning system allows you to make individual warnings be considered as errors, or disable the warning in question in the project settings.

> But users of Godot can just choose not to update and continue using an older version for as long as they want to, so breaking changes don't actually hurt existing Godot developers unless they want access to some other new feature.

While this is true to a certain extent, it should be remembered that [Godot releases are supported for a limited time only](https://docs.godotengine.org/en/latest/about/release_policy.html#release-support-timeline). If you want to get security or platform support updates, you will eventually need to upgrade *at some point*, particularly for games that get regular updates themselves or are networked in some way.

**Comment by dalexeev** [comment]
I agree with you, Godot core and GDScript do not have a consistent thought-out design regarding "insignificant" details. However, I'm not sure how we could improve the situation.

1. A major compatibility breakage is not possible until Godot 5.0. A minor one may be acceptable, but it may not be enough to solve the problem.
2. Two GDScript variants inside Godot 4.x, project settings, or per script directives would make GDScript harder to maintain.
3. We currently have no runtime warnings, and the analyzer cannot always infer types due to the dynamic nature of GDScript.
4. Introducing alternative operators/functions and deprecating existing ones solves the problem to some extent, but it is less convenient from the user perspective.

In addition, all of these options, to varying degrees, make it difficult to reuse code between projects, complicate documentation, tutorials, and user support in community channels.

**Reply by KingSupernova31** [reply]
Oh I certainly agree that breaking changes shouldn't happen until 5.0. I meant this as a longer term suggestion. I expect it would take a long time to design a robust new system anyway; feedback should be solicited from the community on how the new system would work, so that after the redesign happens there'd be no further need to overhaul the same things.

I agree that having a redesign be opt-in by default (like Javascript does with `use_strict`) is inelegant and harder to maintain, but I think such a redesign is sorely needed, and the old version could be phased out over the course of several updates.

I disagree that all of these details are insignificant. I ran into several of these issues naturally within just a few days of starting my first Godot project. Comparing dynamic variables for equality for example is a very common thing to need to do, and having to use `is_same()` every time adds up in typing and mental complexity.

The obstacle to code reuse is a real downside; I just think it's worth it. The tutorials don't seem like a problem to me, since anyone who's just getting into Godot and needs a tutorial would want to use the newest version of GDscript, which means that tutorials don't need to support the old one.
